backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,336 @@
+        } finally {
+          backendAssignmentLock.unlock();
+        }
+      }
+    }, result.completer());
+
+    return result;
+  }
+
+  /**
+   * Finds an available backend from the queue. De-queues all backends which have become defunct since they were added
+   * @return available backend or null if none found
+   */
+  private BackendDetail getAvailableBackendFromPriorityQueue() {
+    BackendDetail availableBackend;
+    while((availableBackend = availableBackendsByPriority.poll()) != null) {
+      if(!availableBackend.isDefunct()) {
+        return availableBackend;
+      }
+    }
+    return (availableBackend == null || availableBackend.isDefunct()) ? null : availableBackend;
+  }
+
+  /**
+   * Associates a backend with process group. Writes to in-memory store and zookeeper
+   * NOTE: Ensure that the backend detail instance being modified has been removed from available backend queue prior to calling this method
+   * @param backendDetail
+   * @param processGroup
+   * @throws Exception
+   */
+  private void associateBackendWithProcessGroup(BackendDetail backendDetail, Recorder.ProcessGroup processGroup)
+      throws Exception {
+    Set<Recorder.ProcessGroup> associatedProcessGroups = new HashSet<>(backendDetail.getAssociatedProcessGroups());
+    associatedProcessGroups.add(processGroup);
+
+    String zNodePath = getZNodePathForBackend(backendDetail.getBackendIPAddress());
+    ZookeeperUtil.writeZNode(curatorClient, zNodePath, BackendDetail.serializeProcessGroups(associatedProcessGroups), false);
+
+    backendDetail.associateProcessGroup(processGroup);
+    processGroupToBackendLookup.put(processGroup, backendDetail.getBackendIPAddress());
+  }
+
+  /**
+   * De associates a backend from process group. Writes to in-memory store and zookeeper
+   * NOTE: Ensure that the backend detail instance being modified has been removed from available backend queue prior to calling this method
+   * @param backendDetail
+   * @param processGroup
+   * @throws Exception
+   */
+  private void deAssociateBackendWithProcessGroup(BackendDetail backendDetail, Recorder.ProcessGroup processGroup)
+      throws Exception {
+    Set<Recorder.ProcessGroup> associatedProcessGroups = new HashSet<>(backendDetail.getAssociatedProcessGroups());
+    associatedProcessGroups.remove(processGroup);
+
+    String zNodePath = getZNodePathForBackend(backendDetail.getBackendIPAddress());
+    ZookeeperUtil.writeZNode(curatorClient, zNodePath, BackendDetail.serializeProcessGroups(associatedProcessGroups), false);
+
+    backendDetail.deAssociateProcessGroup(processGroup);
+    processGroupToBackendLookup.remove(processGroup);
+  }
+
+  private List<BackendDetail> loadDataFromZookeeper() throws Exception {
+    List<BackendDetail> backends = new ArrayList<>();
+    List<String> backendIPAddresses = curatorClient.getChildren().forPath(backendAssociationPath);
```

Comment:
This is called in constructor only. Maybe I can remove the function, and add it directly to constructor if putting this in a separate method can lead to bad usage in future